javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -0,0 +1,853 @@
+        return new PriorityQueue<>(this.comparator, forest);
+    }
+
+    /**
+     * Returns the empty PriorityQueue.
+     *
+     * @param <T> Component type
+     * @return The empty PriorityQueue.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends Comparable<T>> PriorityQueue<T> empty() {
+        return empty(naturalComparator());
+    }
+
+    public static <T> PriorityQueue<T> empty(Comparator<T> comparator) {
+        return new PriorityQueue<>(comparator, List.empty());
+    }
+
+    /**
+     * Narrows a widened {@code PriorityQueue<? extends T>} to {@code PriorityQueue<T>}
+     * by performing a type safe-cast. This is eligible because immutable/read-only
+     * collections are covariant.
+     *
+     * @param queue An {@code PriorityQueue}.
+     * @param <T>   Component type of the {@code PriorityQueue}.
+     * @return the given {@code PriorityQueue} instance as narrowed type {@code PriorityQueue<T>}.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> PriorityQueue<T> narrow(PriorityQueue<? extends T> queue) {
+        return (PriorityQueue<T>) queue;
+    }
+
+
+    @SuppressWarnings("unchecked")
+    public static <T extends Comparable<T>> PriorityQueue<T> of(T... elements) {
+        return of(naturalComparator(), elements);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T extends Comparable<T>> PriorityQueue<T> of(Comparator<T> comparator, T... elements) {
+        return ofAll(comparator, List.of(elements));
+    }
+
+    public static <T extends Comparable<T>> PriorityQueue<T> ofAll(Iterable<? extends T> elements) {
+        return ofAll(naturalComparator(), elements);
+    }
+
+    public static <T> PriorityQueue<T> ofAll(Comparator<T> comparator, Iterable<? extends T> elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        return List.ofAll(elements).foldLeft(empty(comparator), PriorityQueue::insert);
+    }
+
+    @Override
+    public List<T> toList() {
+        List<T> results = List.empty();
+        for (PriorityQueue<T> queue = this; !queue.isEmpty(); queue = queue.deleteMin()) {
+            results = results.prepend(queue.findMin());
+        }
+        return results.reverse();
+    }
+
+    /**
+     * Removes an element from this Queue.
+     *
+     * @return a tuple containing the first element and the remaining elements of this Queue
+     * @throws NoSuchElementException if this Queue is empty
+     */
+    @Override
+    Tuple2<T, PriorityQueue<T>> dequeue() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Enqueues a new element.
+     *
+     * @param element The new element
+     * @return a new {@code Queue} instance, containing the new element
+     */
+    @Override
+    public PriorityQueue<T> enqueue(T element) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public LinearSeq<T> append(T element) {
```

Comment:
all methods like this should return <code>PriorityQueue&lt;T&gt;</code> for properly chaining